### DISCUSSION TITLE
configure strongswan : error: OpenSSL header openssl/evp.h not found!

### DISCUSSION THREAD

**Discussion by AmieeChen** [strongswan/strongswan#311] [Q&A] [answered]
I tried to use MingGW and openssl-1.0.2u to compile strongswan-5.2.0 directly on windows. I used to cross-compile on ubuntu, but I encountered a problem that I couldn't solve. Here is my command to compile the strongswan. 

CC=x86_64-w64-mingw32-gcc CFLAGS="-g -O2 -Wall -Wno-pointer-sign -Wno-format-security -Wno-format -mno-ms-bitfields -I/D/usr/local/ssl/include/openssl/" LDFLAGS="-L/D/usr/local/ssl/lib/" ./configure --disable-defaults --enable-monolithic --enable-static --enable-svc --enable-ikev2 --enable-ikev1 --enable-openssl --enable-nonce --enable-pem --enable-pkcs1 --enable-x509  --enable-socket-win --enable-kernel-wfp --enable-kernel-iph --enable-pubkey --enable-swanctl --with-swanctldir=swanctl --with-strongswan-conf=strongswan.conf

The error information are as follows:

checking clang... no
checking for main in -leay32... yes
checking openssl/evp.h usability... no
checking openssl/evp.h presence... no
checking for openssl/evp.h... no
configure: error: OpenSSL header openssl/evp.h not found!

The config.log in strongswan-5.2.0:

configure:18474: result: no
configure:19706: checking for main in -leay32
configure:19725: x86_64-w64-mingw32-gcc -o conftest.exe -g -O2 -Wall -Wno-pointer-sign -Wno-format-security -Wno-format -mno-ms-bitfields -I/D/usr/local/ssl/include/openssl/  -L/D/usr/local/ssl/lib/ conftest.c -leay32   >&5
configure:19725: $? = 0
configure:19735: result: yes
configure:19743: checking openssl/evp.h usability
configure:19743: x86_64-w64-mingw32-gcc -c -g -O2 -Wall -Wno-pointer-sign -Wno-format-security -Wno-format -mno-ms-bitfields -I/D/usr/local/ssl/include/openssl/  conftest.c >&5
conftest.c:67:25: fatal error: openssl/evp.h: No such file or directory
 #include <openssl/evp.h>
                         ^
compilation terminated.

I am sure that the evp.h is in my path: D:/usr/local/ssl/include/openssl/. But it still can not find it, and I do not know why. And I checked the conftest.c, “#include <openssl/evp.h> “ is not found in the file. I do not know how to solve this. Could anyone give me some suggestions?

**Comment by Thermi** [comment]
So evp.h is at `/D/usr/local/ssl/include/openssl/openssl/evp.h`? ;)

**Reply by AmieeChen** [reply]
yeap, usr/local/ssl under my drive D is the default path for compiling openssl.

**Reply by AmieeChen** [reply]
And evp.h is at "D:/usr/local/ssl/include/openssl/evp.h"

**Comment by Thermi** [comment, marked as answer]
Note the double openssl in the path I constructed from the argument and include line.

Am April 10, 2021 7:33:32 AM UTC schrieb AmieeChen ***@***.***>:
>And evp.h is at "D:/usr/local/ssl/include/openssl/evp.h"
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/strongswan/strongswan/discussions/311#discussioncomment-593329

-- 
Sent from mobile

**Reply by AmieeChen** [reply]
I noted that and reconfired my openssl path. But there is only one  openssl in my path.

**Reply by Thermi** [reply]
These are your passed CFLAGS:
```
CC=x86_64-w64-mingw32-gcc CFLAGS="-g -O2 -Wall -Wno-pointer-sign -Wno-format-security -Wno-format -mno-ms-bitfields -I/D/usr/local/ssl/include/openssl/" LDFLAGS="-L/D/usr/local/ssl/lib/" ./configure --disable-defaults --enable-monolithic --enable-static --enable-svc --enable-ikev2 --enable-ikev1 --enable-openssl --enable-nonce --enable-pem --enable-pkcs1 --enable-x509 --enable-socket-win --enable-kernel-wfp --enable-kernel-iph --enable-pubkey --enable-swanctl --with-swanctldir=swanctl --with-strongswan-conf=strongswan.conf
```

Note the specified include path `/D/usr/local/ssl/include/openssl/`.
In the code, the include is specified as `#include <openssl/evp.h>`. The C compiler resolves the include, due to the usage of <>, as dependent on the include paths. So it looks, at `/D/usr/local/ssl/include/openssl/` + `openssl/evp`. That is then `/D/usr/local/ssl/include/openssl/openssl/evp.h`. Your openssl headers are at `/D/usr/local/ssl/include/openssl/` exactly though. So you need to get rid of that extra `openssl` in the path. The straight forward solution to that is to just remove the `openssl/` from `/D/usr/local/ssl/include/openssl/`. That is then `/D/usr/local/ssl/include/`. Together with `openssl/evp.h`, that then results in the correct path `/D/usr/local/ssl/include/openssl/evp.h`. So use `-I/D/usr/local/ssl/include/` instead of `-I/D/usr/local/ssl/include/openssl/`

**Reply by AmieeChen** [reply]
Wow, It really works! Thank you so much.

**Reply by Thermi** [reply]
yw.

**Reply by Thermi** [reply]
Would you be so kind to select the answer as correct?
